src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs

```diff
@@ -48,32 +49,24 @@ public class UsersService : UsersServiceBase, IUsersService
             _authService = EnsureArg.IsNotNull(authService, nameof(authService));
             _queueService = EnsureArg.IsNotNull(queueService, nameof(queueService));
             _googleFitTokensService = EnsureArg.IsNotNull(googleFitTokensService, nameof(googleFitTokensService));
+            _authStateService = EnsureArg.IsNotNull(authStateService, nameof(authStateService));
             _utcNowFunc = EnsureArg.IsNotNull(utcNowFunc);
             _logger = logger;
         }
 
         public override string PlatformName => GoogleFitConstants.GoogleFitPlatformName;
 
         /// <inheritdoc/>
-        public async Task ProcessAuthorizationCallback(string authCode, string state, CancellationToken cancellationToken)
+        public async Task ProcessAuthorizationCallback(string authCode, string nonce, CancellationToken cancellationToken)
         {
             if (string.IsNullOrWhiteSpace(authCode))
             {
                 // This exception message will be visible to the caller.
                 throw new Exception("The Google authorization service failed to return an access code.");
             }
 
-            AuthState authState;
-
-            try
-            {
-                authState = AuthState.Parse(state);
-            }
-            catch (Exception e) when (e is ArgumentException || e is JsonSerializationException)
-            {
-                // This exception message will be visible to the caller.
-                throw new Exception("The Google authorization service failed to return the expected authorization state.");
-            }
+            var validNonce = EnsureArg.IsNotNullOrWhiteSpace(nonce);
+            AuthState authState = await _authStateService.RetrieveAuthState(validNonce, cancellationToken);
```

Comment:
I think we need to catch any Exception that might happen when we try to retrieve the AuthState using the nonce and throw an Exception with an error that is appropriate for the general public. You can see in the error you deleted above, we are trying to make sure that public facing errors are more polished.

Reply:
Yes, in `RetrieveAuthState` I catch any exception and throw.  This is ultimately caught again in `GoogleFitAuthorizationHandler.Callback`, which returns a 404.  

I deleted the exception shown here, because that message is no longer accurate 100% of the time.  Now there could be a failure either on Google's end with the returned nonce, or in the retrieval of the blob.  

I am open to suggestions on what the exception message should be.  I kept it general for now, because I didn't know of a better way to phrase it ("Failed to retrieve the auth state.")

Reply:
I think it would be fine to just say "An error occurred while validating the Google authorization callback 'state' parameter."